frontend/catalyst/jax_tracer.py

```diff
@@ -541,47 +664,68 @@ def trace_quantum_function(
-
-            out_classical_tracers = [trace.full_raise(t) for t in out_classical_tracers]
-
-            jaxpr, out_type, consts = ctx.frames[trace].to_jaxpr2(
-                out_classical_tracers + out_quantum_tracers
+            return_values_flat, return_values_tree = jax.tree_util.tree_flatten(
+                return_values, is_leaf=is_leaf
             )
-            jaxpr._outvars = jaxpr._outvars[:-1]  # pylint: disable=protected-access
-            out_type = out_type[:-1]
-            # TODO: `check_jaxpr` complains about the `AbstractQreg` type. Consider fixing.
-            # check_jaxpr(jaxpr)
 
-    closed_jaxpr = ClosedJaxpr(jaxpr, consts)
-    out_avals, _ = unzip2(out_type)
+            tapes, post_processing = apply_transform(qnode, quantum_tape, return_values_flat)
 
-    abstract_results = tree_unflatten(
-        out_classical_tree, [ShapeDtypeStruct(a.shape, a.dtype, a.named_shape) for a in out_avals]
-    )
-    return closed_jaxpr, abstract_results
+        # (2) - Quantum tracing
+        results_tracers, results_abstract = [], []
+        is_program_transformed = qnode and qnode.transform_program
+        for tape in tapes:
+            # If the program is batched, that means that it was transformed.
+            # If it was transformed, that means that the program might have
+            # changed the output. See `split_non_commuting`
+            if is_program_transformed:
+                # If the program is transformed
+                # trees == out_classical_tree in the next line section.
+                output = tape.measurements
+                _, trees = jax.tree_util.tree_flatten(output, is_leaf=is_leaf)
+            else:
+                output = return_values_flat
+                trees = return_values_tree
+
+            with EvaluationContext.frame_tracing_context(ctx, trace):
+                qdevice_p.bind(spec="kwargs", val=str(device.backend_kwargs))
+                qdevice_p.bind(spec="backend", val=device.backend_name)
+                qreg_in = qalloc_p.bind(len(device.wires))
+                qrp_out = trace_quantum_tape(tape, device, qreg_in, ctx, trace)
+                meas, meas_trees = trace_quantum_measurements(device, qrp_out, output, trees)
+                out_quantum_tracers = [qrp_out.actualize()]
+                qdealloc_p.bind(qreg_in)
+
+                tracers = [trace.full_raise(m) for m in meas]
+                results_tracers += tracers
+
+                jaxpr, out_type, _ = ctx.frames[trace].to_jaxpr2(tracers + out_quantum_tracers)
+                jaxpr._outvars = jaxpr._outvars[:-1]  # pylint: disable=protected-access
+                out_type = out_type[:-1]
+
+                out_avals, _ = unzip2(out_type)
+                abstract_results = tree_unflatten(
+                    meas_trees,
+                    [ShapeDtypeStruct(a.shape, a.dtype, a.named_shape) for a in out_avals],
+                )
+                # This mimics the return type from qnodes.
```

Comment:
Yes similar to the message above we could relax this constraint in the future.

Reply:
Added comments here: https://github.com/PennyLaneAI/catalyst/pull/280/commits/74d4baae5a8c8e92586f976c506c8968a9f5e137